### PR TITLE
Security

### PR DESCRIPTION
**Pull request by hujw77**
- [x] Cross chain filter support
- [x] `MessageRoot` support multi-channel
- [x] Message struct add inbound channel address to distinguish messages from other channels
- [x] outbound channel add whitlist 
- [x] test and doc
- [x] `MessageRoot` support multi-chain
- [x] Add guard system

### CODE REVIEW

File: contracts/bridge/contracts/binance/BasicInboundChannel.sol [hunk 1]
```diff
@@ -83,18 +85,28 @@ contract BasicInboundChannel {
...
             nonce = nonce + 1;
 
-            // Deliver the message to the target
-            (bool success, ) =
-                messages[i].target.call{value: 0, gas: MAX_GAS_PER_MESSAGE}(
-                    messages[i].payload
-                );
-
-            emit MessageDispatched(messages[i].nonce, success);
+            try ICrossChainFilter(message.target).filtered(message.origin)
```
Comment (HackFisher) [inline contracts/bridge/contracts/binance/BasicInboundChannel.sol:94]: What if some else contract also implement filter() interface by accident.

Comment (hujw77) [reply contracts/bridge/contracts/binance/BasicInboundChannel.sol]: I changed the interface to a less collision-prone name, and the registration pattern requires more complex operations on the part of the app layer.

File: contracts/bridge/contracts/binance/BasicInboundChannel.sol [hunk 2]
```diff
@@ -27,18 +54,73 @@ contract BasicInboundChannel {
...
+     * @param nonce The message nonce
+     * @param result The message result
+     * @param returndata The return data of message call, when return false, it's the reason of the error
+     */
+    event MessageDispatched(uint256 indexed nonce, bool indexed result, bytes returndata);
+
+    /* State */
+
+    /**
+     * @dev The position of the leaf in the chain message merkle tree, index starting with 0
```
Comment (HackFisher) [inline contracts/bridge/contracts/binance/BasicInboundChannel.sol:68]: The document does not match

File: contracts/bridge/contracts/binance/BasicInboundChannel.sol [hunk 3]
```diff
@@ -27,18 +54,73 @@ contract BasicInboundChannel {
...
+     */
+    uint256 public chainId;
 
-    uint64 public nonce;
+    /**
+     * @dev The position of the leaf in the lane message merkle tree, index starting with 0
```
Comment (HackFisher) [inline contracts/bridge/contracts/binance/BasicInboundChannel.sol:73]: The document does not match

File: contracts/bridge/contracts/binance/BasicOutboundChannel.sol [hunk 4]
```diff
@@ -2,23 +2,37 @@
...
 
+    constructor() public {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
     /**
      * @dev Sends a message across the channel
      */
     function submit(bytes calldata payload) external override {
+        require(hasRole(OUTBOUND_ROLE, msg.sender), "Channel: not-authorized");
```
Comment (HackFisher) [inline contracts/bridge/contracts/binance/BasicOutboundChannel.sol:35]: Missing docs, and 1. why it requires auth here? 2. how the fees are paid?

Comment (hujw77) [reply contracts/bridge/contracts/binance/BasicOutboundChannel.sol:35]: Related to the task `outbound channel add whitlist`.
At the beginning of the bridge, only whitelisted apps may be required to send messages from Ethereum to Darwinia.

File: contracts/bridge/contracts/binance/GuardRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+contract GuardRegistry {
+    /* Events */
+
+    event GuardRegistryUpdated(uint256 id, uint256 len, bytes32 root);
+
+    /* State */
+
+    uint256 public guardSetId;
+    uint256 public numOfGuards;
+    bytes32 public guardSetRoot;
```
[the remark this answers is not in the document]
Comment (hujw77) [reply contracts/bridge/contracts/binance/GuardRegistry.sol:29]: https://www.notion.so/itering/BEEFY-Commitment-7f705fe6dade482d8119a10b441d1bb1#5780960dd4d14295b6467d89c0291b78

File: contracts/bridge/contracts/binance/BasicInboundChannel.sol (hunk 2 above)
Comment (hujw77) [reply contracts/bridge/contracts/binance/BasicInboundChannel.sol:68]: https://www.notion.so/itering/Basic-Message-Channel-c41f0c9e453c478abb68e93f6a067c52#9607848fcd674196809f060470642707

**Review by HackFisher — APPROVED** [review]

**Pull request merged**